lib/autokey/service.py

```diff
@@ -491,6 +498,20 @@ def execute(self, script: model.Script, buffer=''):
 
         self.mediator.send_string(trigger_character)
 
+    @threaded
+    def execute_path(self, path: pathlib.Path, buffer=''):
+        logger.debug("Script runner executing: {}".format(path))
+        scope = self.scope.copy()
+        # scope["store"] = script.store
+        # Overwrite __file__ to contain the path to the user script instead of the path to this service.py file.
+        scope["__file__"] = path.resolve()
+        try:
+            exec(open(path).read(), scope)
```

Comment:
This open() leaks an opened file handle, which is never closed.
Using `path.read_text()` automatically closes the file after reading the content, as indicated by
https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_text

Reply:
Interesting! I note that it was only added in 3.5. I don't know what you consider a minimal supported python 3 to be.

I'll change it. Teaches me to blindly copy the SO solution.

Reply:
Oh, actually it wasn't SO, it was an equivalent function somewhere else in autokey. So I'll change it twice.

Reply:
> Oh, actually it wasn't SO, it was an equivalent function somewhere else in autokey. So I'll change it twice.

Huh? OK then, TIL. Go ahead ;)

> Interesting! I note that it was only added in 3.5. I don't know what you consider a minimal supported python 3 to be.

Huh? (The second…) The Readme used to be stating that Python 3.5 is the minimal supported Python version.

Reply:
Oops, sorry. Should have checked the readme.

I'll see if I can find that function, I might me misremembering.

Reply:
No no no. Sorry from my side. It seems that this bit vanished when the dependency information was de-duplicated. There used to be a slightly outdated section in the README and in the Wiki. Both got merged into the Wiki. But it seems that the minimal Python version was not properly carried to the Wiki page.

Reply:
I take it back, it wasn't already there. Normally there was no need to read the code in a script, I added it as part of arbitrary scripts. But I think there are two instances, I'll change both.

Statement about SO stands.

Reply:
Also, regarding the wiki: As I make changes, they change relevancy bits of wiki documentation. But it seems the wiki can only apply to the current latest release. Any suggestion for storing up notes on what to update in the wiki? Maybe a file somewhere with notes

Reply:
Oh, And about:

> Teaches me to blindly copy the SO solution.

You are sure that the license of the copied SO solution was GPLv3 compatible? :smiling_imp: 

(This, plus possible issues in the actual code are the two major obstacles when looking up stuff there…)



> Also, regarding the wiki: As I make changes, they change relevancy bits of wiki documentation. […]

I think, handling this with some additional notes at the relevant bits may be sufficient. For example, noting at which version functions were added and changed. For example, somewhat like this: https://documen.tician.de/pyopencl/runtime_queue.html#pyopencl.Event.from_int_ptr

That documentation has statements like these, to indicate changes:

> New in version 2013.2.
> 
> Changed in version 2016.1: retain added.

Reply:
Ok. So if I make a change in my branch, I hope it will eventually become a PR, which will eventually be merged into dev, which will become `0.96`. What stage to I say in the wiki "Changed in `0.96`"?

Reply:
I’d say some time prior to releasing the next version. The changelog has to be written anyways, so that time frame looks like it may be a good idea. I’m not totally sure either.
The best way may be actually placing it into the docstrings and automatically generate the relevant pages. (Related to #330)

Reply:
I wholeheartedly agree with the last part, but for macros that could be problematic because they aren't associated with a particular function to create a docstring. Unless we
just chuck the docstring in `macro.py`

Reply:
This has been fixed.